### PR TITLE
lnwallet/test: add coverage for evaluateHtlcView

### PR DESCRIPTION
**Pull request by carlaKC**
This PR extracts relevant unit tests in #4093 so that we can increase coverage in this area, even if we decide not to go ahead with the refactor in the referenced PR.

### CODE REVIEW

File: lnwallet/channel.go [hunk 1]
```diff
@@ -2617,6 +2569,64 @@ func (lc *LightningChannel) evaluateHTLCView(view *htlcView, ourBalance,
...
+		}
+
+		addEntry := updateLog.lookupHtlc(entry.ParentIndex)
+
+		switch {
+		// We check if the parent entry is not found at this point.
+		// This could happen for old versions of lnd, and we return an
+		// error to gracefully shut down the state machine if such an
+		// entry is still in the logs.
+		case addEntry == nil:
```
Comment (cfromknecht) [inline lnwallet/channel.go:2603]: these three cases are still uncovered, do we want to test them in this pr?

Comment (carlaKC) [reply lnwallet/channel.go:2627]: we do indeed! Added them in a separate fetchParent test bc it's easier to split up that way.

Comment (cfromknecht) [reply lnwallet/channel.go:2627]: awesome, nice addition! looking at coveralls the middle case `case remoteChain && addEntry.addCommitHeightRemote == 0:` is still uncovered, maybe just a configuration issue in the unit test?

Comment (carlaKC) [reply lnwallet/channel.go:2627]: Local and remote log values mixed up in the `remote log, remote add height 0` case, fixed :)

File: lnwallet/channel_test.go [hunk 2]
```diff
@@ -8193,3 +8193,622 @@ func checkExpectedHtlcs(t *testing.T, actual []*PaymentDescriptor,
 		}
 	}
 }
+
+// heights represents the heights on a payment descriptor.
```
Comment (joostjager) [inline lnwallet/channel_test.go:8197]: type commit message

Comment (joostjager) [inline lnwallet/channel_test.go:8197]: typo commit message

File: lnwallet/channel_test.go [hunk 3]
```diff
@@ -7812,3 +7811,385 @@ func TestFetchParent(t *testing.T) {
...
+		// theirFeeUpdatePerSat is the fee rate *in satoshis* that we
+		// expect if we update to ourFeeUpdateAmt.
+		theirFeeUpdatePerSat = chainfee.SatPerKWeight(10)
+	)
+
+	// fetchParentFromMap takes a map of htlc index to htlc parent indexes
+	// and returns a fetch parent function which returns a payment
+	// descriptor with that parent index set as the htlc add index. If the
+	// child htlc is not a key in the map, this function will error.
+	fetchParentFromMap := func(parentMap map[uint64]uint64) fetchParent {
```
Comment (joostjager) [inline lnwallet/channel_test.go:7851]: This looks risky to me. To replace such a low-level component of the system under test. Isn't it testable using the real `fetchParent`?

File: lnwallet/channel_test.go [hunk 4]
```diff
@@ -7812,3 +7811,385 @@ func TestFetchParent(t *testing.T) {
...
+					TotalMSatSent:     0,
+					TotalMSatReceived: 0,
+				},
+			}
+
+			view := &htlcView{
+				ourUpdates:   test.ourHtlcs,
+				theirUpdates: test.theirHtlcs,
+				feePerKw:     feePerKw,
+			}
```
Comment (joostjager) [inline lnwallet/channel_test.go:8127]: I am wondering, is this indeed the right abstraction level for this test? I can see that the goal is to test `evaluteHtlcView` in isolation, but perhaps it is better to test at a slightly higher level to be sure that the interaction between `fetchHTLCView` and `evaluateHtlcView` is captured as well.

Comment (carlaKC) [reply lnwallet/channel_test.go:8127]: I think there's too much going on in `evaluateHtlcView` to achieve this level of coverage at a higher level. There's a lot of subtlety with state mutation, which log we use etc. If anything, another test for `computeView` would be nice.

File: lnwallet/channel_test.go (hunk 3 above)
Comment (carlaKC) [reply lnwallet/channel_test.go]: Good point, replaced with the real thing.

**Comment by carlaKC** [comment]
> But because this PR adds white-box style unit tests, it may become even less attractive to move in that direction. It isn't possible anymore to freely move the internals around. Would be interested to hear some opinions about this.

My opinion on this would be that these tests make it easier to change this code, in a way. When I wanted to change some things around in this part of the codebase, these tests were my first step to make sure that I could pick up any behaviour changes that the refactor caused. If we are going to take the time to refactor lnwallet, a very serious change, then a unit test really shouldn't be a blocker imo.

File: lnwallet/channel.go [hunk 5]
```diff
@@ -2641,6 +2590,57 @@ func (lc *LightningChannel) evaluateHTLCView(view *htlcView, ourBalance,
...
+		logName   string
+	)
+
+	if remoteLog {
+		updateLog = lc.remoteUpdateLog
+		logName = "remote"
+	} else {
+		updateLog = lc.localUpdateLog
+		logName = "local"
+	}
```
[the remark this answers is not in the document]
Comment (carlaKC) [reply lnwallet/channel.go:2609]: Not sure if saving a few lines in a unit test is worth this change though?

File: lnwallet/channel_test.go [hunk 6]
```diff
@@ -7865,3 +7864,392 @@ func TestFetchParent(t *testing.T) {
...
+				},
+
+				// Create update logs for local and remote.
+				localUpdateLog:  newUpdateLog(0, 0),
+				remoteUpdateLog: newUpdateLog(0, 0),
+			}
+
+			for _, htlc := range test.ourHtlcs {
+				if htlc.EntryType == Add {
+					lc.localUpdateLog.appendHtlc(htlc)
```
[the remark this answers is not in the document]
Comment (carlaKC) [reply lnwallet/channel_test.go:8170]: Looked at this, it blows up the setup required for the test because we need sane commitments and all sorts.

**Comment by Roasbeef** [comment]
Needs rebase.

**Pull request merged**